### PR TITLE
feat: Add Bitbucket integration V1 PUL-1364

### PR DESCRIPTION
**Pull request by claudiacarpinteiro**
Pulse will now support one-click integration with Bitbucket. For V1, Pulse users will be able to install the Bitbucket integration connected to a single workspace and see their real-time accelerate metrics via pull request data.

### :eyes: Live preview
https://feat-bitbucket-integration-pul-1364--docs-pulse.netlify.app/one-click-integrations/bitbucket-integration/
https://feat-bitbucket-integration-pul-1364--docs-pulse.netlify.app/#2-pushing-data-to-pulse
https://feat-bitbucket-integration-pul-1364--docs-pulse.netlify.app/metrics/accelerate/#see-also

### :construction: To do
-   [x] Add screenshots for the installation flow
-   [x] Update warnings for feature limitations
-   [x] Update Bitbucket permissions
-   [x] Update screenshots with the final version
-   [x] Wait for feature release

### CODE REVIEW

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [inline docs/one-click-integrations/bitbucket-integration.md:6]: @mrfyda Does Pulse integrate with Bitbucket Cloud only? Or also with Bitbucket Server?

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [inline docs/one-click-integrations/bitbucket-integration.md:22]: @afsmeira The Pulse Bitbucket App is installed in a Bitbucket account? (because we only ask for the workspace after the Bitbucket App is installed, right?...)

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [inline docs/one-click-integrations/bitbucket-integration.md:8]: For V1, we’ll only calculate the following accelerate metrics in real time:

- Deployment frequency
- Lead time for changes

All the remaining metrics will be delivered for V3:

- Changes metrics (Time to open, Time to merge)
-  Reviews metrics (Time to first review, Time to first approval, Time from first to last approval, Time from last approval to merge, Outlier review details)
- Work in progress

@mrfyda Can you please confirm?

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [inline docs/one-click-integrations/bitbucket-integration.md:48]: @afsmeira Can you please confirm if this information is correct for Bitbucket integration? I based on the GitHub content and did some adjustments (for example, replaced rebase with fast-forward).

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [inline docs/one-click-integrations/bitbucket-integration.md:50]: @afsmeira Can you please validate the collected data for Bitbucket integration? I based it on the GitHub content, but maybe the mapping to the Pulse data model needs some tweaks.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [inline docs/one-click-integrations/bitbucket-integration.md:105]: @afsmeira I will complete this section with the information you'll provide in the scope of [PUL-1507](https://codacy.atlassian.net/browse/PUL-1507).
If you prefer to provide that information here instead of creating a new internal document, please go ahead. What's easier for you.

File: docs/metrics/accelerate-changes.md
Comment (claudiacarpinteiro) [inline docs/metrics/accelerate-changes.md:3]: @mrfyda @claudiacodacy Also, I believe we should include a similar business-oriented message about the current Bitbucket integration limitations on the pages describing the metrics that won't be delivered for V1. Do you agree?

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacodacy) [reply docs/one-click-integrations/bitbucket-integration.md:8]: that's right clau! :)

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacodacy) [reply docs/one-click-integrations/bitbucket-integration.md:4]: I think a message like this could work well! :) I'd only tweak the tone a bit to sound more friendly, for example, "Pulse's Bitbucket integration is new and still in beta, therefore some metrics may not be available just yet. If you'd like to know more about what's coming next, feel free to contact us" <- not sure if here we can use this to open the intercom directly for example 💭 tweak it as you think it's best 🙏 !

File: docs/metrics/accelerate-changes.md
Comment (claudiacodacy) [reply docs/metrics/accelerate-changes.md:3]: That makes sense to me! :) we should even consider this inside the platform for v2, perhaps... 🤔

File: docs/index.md
Comment (mrfyda) [inline docs/index.md:40]:
```suggestion
    -   **[GitHub integration](one-click-integrations/github-integration.md)** or **[Bitbucket integration](one-click-integrations/bitbucket-integration.md)**: reports changes and deployments
```

change to `or` because while we're not going to block users from configuring both git providers, it's not a use case that we planned for

File: docs/one-click-integrations/bitbucket-integration.md
Comment (mrfyda) [reply docs/one-click-integrations/bitbucket-integration.md:4]: 1. we can link to the Accelerate metrics mentioned they will be the ones supported
2. highlight that historical import won't be available
2. highlight that Deployments will only be detected thought pull requests

"Pulse's Bitbucket integration is new and still in Beta. For now, only the [Accelerate metrics] are available, with Deployments only being detected from Pull requests.
If at first you don't see any data, don't worry. At the moment, we're only collecting new data, but we'll automatically import historical data and backfill your charts within a few weeks."

File: docs/one-click-integrations/bitbucket-integration.md
Comment (mrfyda) [reply docs/one-click-integrations/bitbucket-integration.md:22]: yes, it's associated with a BB user account
it gives access to what the person sees and to impersonate them in Bitbucket

that means we'll only be able to see the workspaces they have access to, and within the workspace only have visibility over what they can already see in BB
for that reason, we're only allowing users to integrate with workspaces they have Admin rights over (and therefore have complete visibility over)

File: docs/one-click-integrations/bitbucket-integration.md
Comment (mrfyda) [inline docs/one-click-integrations/bitbucket-integration.md:48]:
```suggestion
-   Pulse associates all Bitbucket user groups of the author of a merged pull request with the corresponding deployment, excluding user groups with less than two members. Pulse only takes changes to Bitbucket teams into account on pull requests merged after those changes.
```
I believe BB concept of teams is called user groups. @claudiacodacy can confirm that
I'm not sure which one you'd like to use here

Comment (claudiacarpinteiro) [reply docs/one-click-integrations/bitbucket-integration.md:48]: Yep, you're right. It seems that in Bitbucket you can have [2 types of accounts: Individuals and Team accounts](https://community.atlassian.com/t5/Bitbucket-questions/Difference-between-teams-and-groups/qaq-p/755346), but they use [User groups to organize members of a workspace](https://support.atlassian.com/bitbucket-cloud/docs/organize-workspace-members-into-groups/). 👍

File: docs/one-click-integrations/bitbucket-integration.md
Comment (afsmeira) [reply docs/one-click-integrations/bitbucket-integration.md:50]: Screenshot 2022-12-14 at 11 59 04

☝️ These are the permissions we request.

- **Email** - to get the email of the user doing the integration.
- **Account** (requires Email) - to list workspaces, user groups and check workspace membership.
- **Workspace membership** - I actually don't think this is required. I will investigate further.
- **Repositories** - get repository default branch, list commits.
- **Pull request** (requires repository) - list pull requests and their activity.
- **Webhooks** - to created and delete workspace webhooks.

Regarding webhooks, we subscribe to:
- Pull request events
    - Created
    - Updated
    - Fulfilled
    - Rejected
    - Approved
    - Unapproved
    - Changes request created
    - Changes request removed
- Repository events
    - Pushed reference (we're only interested tags)
    - Transfered (repository transfered to the selected workspace, so that we can start historical data import)
    - Forked (repository created in the selected workspace as a fork of another repository, so that we can start historical data import)
    
The permissions and events will be requested and subscribed to since V1 but we might not use them all in V1.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (afsmeira) [reply docs/one-click-integrations/bitbucket-integration.md:105]: Check my comment above, I think it's all there.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (afsmeira) [reply docs/one-click-integrations/bitbucket-integration.md]: It is correct 👍

File: docs/one-click-integrations/bitbucket-integration.md
Comment (prcr) [reply docs/one-click-integrations/bitbucket-integration.md:27]: I think it's the first time I see a suggestion from Codacy to fix a Markdown issue. :nerd_face:

File: docs/one-click-integrations/bitbucket-integration.md
Comment (prcr) [inline docs/one-click-integrations/bitbucket-integration.md:43]: I see that we're mentioning the initial limitations of the Bitbucket integration in quite a few places. I suggest using a comment as a small trick to help identify these places, so it's easier to find them all and not miss any when it's time to remove them:

```suggestion
    <!--NOTE Limitation from Bitbucket integration V1-->Pulse will use your merged pull requests to automatically detect your deployments. Other strategies to detect deployments will be available with a future version of Pulse.
```

**Review by prcr — COMMENTED** [review]
It's looking good and quite advanced already! :star_struck: It's also great to see the contributions from the rest of the Pulse team. :heart:

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [reply docs/one-click-integrations/bitbucket-integration.md:105]: Thank you so much for all the details, @afsmeira . 🙏  I updated the permissions table with the provided information.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [reply docs/one-click-integrations/bitbucket-integration.md:50]: Updated permissions table.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [reply docs/one-click-integrations/bitbucket-integration.md]: I updated the message based on your feedback. 👍 
I didn't include the "feel free to contact us" link for now. Let me know if you still want to include this link and what is the target url.

File: docs/metrics/accelerate-changes.md
Comment (claudiacarpinteiro) [reply docs/metrics/accelerate-changes.md]: @prcr I agree with you, let's keep the limitations centrally on the page dedicated to the Bitbucket integration. People integrating with GitHub, for example, and reading the reference for the other metrics, don't really care about Bitbucket limitations.
I removed the messages from review and changes metrics, and work in progress metrics.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [reply docs/one-click-integrations/bitbucket-integration.md]: Thanks! I removed the webhook and tweaked the message according to the UI.

File: docs/index.md
Comment (nicklem) [inline docs/index.md:40]: I wonder if it may make sense to invert the fragments now that we have multiple items per line:
```suggestion
    -   **Report changes and deployments**: [GitHub integration](one-click-integrations/github-integration.md) or [Bitbucket integration](one-click-integrations/bitbucket-integration.md)
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [inline docs/one-click-integrations/bitbucket-integration.md:35]: A fragment is missing here:
```suggestion
1.  Choose the workspace that you want to connect the integration to. You will only be able to select a workspace that you are the owner.
```
Although it reads a bit clunky. We may also want to consider: `Choose the workspace that you want to connect to.` or `Choose the workspace that you want to integrate with.`, depending on which one makes more sense.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [inline docs/one-click-integrations/bitbucket-integration.md:41]: @codacy/techwriters We should decide on verb tenses at some point.

The MS style guide suggests [using to the present tense](https://learn.microsoft.com/en-us/style-guide/grammar/verbs#verb-tense) for readability.

```suggestion
    If there is an error please [contact support](mailto:pulsesupport@codacy.com).
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [inline docs/one-click-integrations/bitbucket-integration.md:46]: Same here. We should default to the present tense. If we use the future tense, users may incorrectly understand that this feature is not yet ready.

```suggestion
    <!--NOTE Limitation from Bitbucket integration V1-V2-->Pulse uses your merged pull requests to automatically detect your deployments. Other strategies to detect deployments will be available with a future version of Pulse.
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [inline docs/one-click-integrations/bitbucket-integration.md:58]: If I understood correctly, I believe rewording the sentence may improve readability. Something like:
```suggestion
-   The metrics that Pulse displays on the dashboards, calculated using the collected data
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [inline docs/one-click-integrations/bitbucket-integration.md:106]: This one is a bit more debatable, but in the spirit of readability and clarity about this being a current feature, we may want to use the present tense here as well.

```suggestion
<sup><span id="deployment-teams">2</span></sup>: Adding a new user group or changing the composition of a user group on Bitbucket only affects new data from that moment onwards and doesn't have an immediate impact on the dashboards. Also, deleted user groups on Bitbucket are still visible in Pulse.
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (claudiacarpinteiro) [reply docs/one-click-integrations/bitbucket-integration.md:35]: I prefer to keep the "owner" to match Bitbucket terminology. Also, I'll use "connect the integration to", to be consistent with the UI (on the UI we have "To connect to a specific workspace you must have owner permissions").

So, to keep the sentence consistent with the UI, I'll go with:
"Choose the workspace that you want to connect the integration to. You can only select a workspace where you have owner permissions."
(I'll do the update manually.)

Comment (claudiacarpinteiro) [reply docs/one-click-integrations/bitbucket-integration.md]: Updated!

File: docs/one-click-integrations/bitbucket-integration.md
Comment (afsmeira) [inline docs/one-click-integrations/bitbucket-integration.md:50]: If 'Squash and merge' is used for Bitbucket, Pulse will not produce accurate accelerate metrics (the only metrics available for V1).

We might be able to overcome this but we don't have time to investigate and implement it for V1, so it will be left for later (see https://codacy.atlassian.net/browse/PUL-1525).

File: docs/one-click-integrations/bitbucket-integration.md
Comment (afsmeira) [inline docs/one-click-integrations/bitbucket-integration.md:149]: This event is trigger with pushing branches, commits, tags to a repository, but we only care about tags. Not sure if that is worth mentioning.

File: docs/one-click-integrations/bitbucket-integration.md
Comment (afsmeira) [inline docs/one-click-integrations/bitbucket-integration.md:86]: This is a bit different from GH. Since there is no unique ID for a pull request in BB, we use the pull request number (unique within a repository) and the timestamp of when the PR is merged.

Not sure if it's worth going into too much detail.

Comment (nicklem) [reply docs/one-click-integrations/bitbucket-integration.md:86]: We may just update the wording to be more generic, such as `unique pull request identifier`

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [reply docs/one-click-integrations/bitbucket-integration.md:149]: Note to self: consider editing as below.

```suggestion
        <li><code>repo:push</code>: pushes of commits, branches, or tags to a repository</li>
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [reply docs/one-click-integrations/bitbucket-integration.md:86]:
```suggestion
            <li><code>deploy_id</code>: unique pull request identifier</li>
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [reply docs/one-click-integrations/bitbucket-integration.md]: See https://github.com/codacy/pulse-user-docs/pull/160#discussion_r1053142215

File: docs/one-click-integrations/bitbucket-integration.md
Comment (afsmeira) [reply docs/one-click-integrations/bitbucket-integration.md:54]: @nicklem It's fixed for V1 -- https://codacy.atlassian.net/browse/PUL-1526

Comment (nicklem) [reply docs/one-click-integrations/bitbucket-integration.md:54]:
```suggestion
!!! note
    <!--NOTE Limitation from Bitbucket integration V1-V2-->If you use a [squash merge strategy](https://support.atlassian.com/bitbucket-cloud/docs/merge-a-pull-request/#Merge-strategies) when merging the pull request, Pulse may produce inaccurate accelerate metrics.
```

**Review by afsmeira — APPROVED** [review]

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [inline docs/one-click-integrations/bitbucket-integration.md:54]:
```suggestion
!!! note
    -   <!--NOTE Limitation from Bitbucket integration V1-V2-->Pulse uses your merged pull requests to automatically detect your deployments. Other strategies to detect deployments will be available with a future version of Pulse.
    -   <!--NOTE Limitation from Bitbucket integration V1-V2-->If you use a [squash merge strategy](https://support.atlassian.com/bitbucket-cloud/docs/merge-a-pull-request/#Merge-strategies) when merging the pull request, Pulse may produce inaccurate accelerate metrics.

-   Pulse considers a deployment every merged pull request that **targets the default branch** of the repository.
-   The deployment date is the timestamp when the corresponding pull request is merged.
-   The set of changes in a deployment is the list of commits in the corresponding pull request.
-   Pulse associates all the Bitbucket user groups of the author of a merged pull request with the corresponding deployment, excluding the user groups with less than two members. Pulse only takes the changes to Bitbucket user groups into account on pull requests merged after those changes.
```

File: docs/one-click-integrations/bitbucket-integration.md
Comment (prcr) [inline docs/one-click-integrations/bitbucket-integration.md:29]:
```suggestion
1.  Choose the workspace that you want to connect the integration to and click **Complete setup** to confirm. You can only select a workspace where you have owner permissions.
```

**Review by prcr — COMMENTED** [review]
It's looking good! :star_struck: 

We're now in the final polishing stages: I left a minor suggestion and recommend cropping the screenshots just a bit "tighter" since that means they're less likely to be shrinked when displayed to users (you can compare them to the GitHub integration ones).

File: docs/one-click-integrations/bitbucket-integration.md
Comment (nicklem) [reply docs/one-click-integrations/bitbucket-integration.md]: OK, thanks! I updated as suggested and cropped the screenshots to match the dimensions of the GitHub ones.

**Review by prcr — APPROVED** [review]
From my point of view, the changes are ready to be deployed! :rocket: 

We should also update the screenshots for the other integrations, as the sidebar has now a new entry for the Bitbucket integration (besides also having an updated "look and feel"). But that can be done in a separate pull request that doesn't block this one. :+1:

File: docs/one-click-integrations/bitbucket-integration.md
Comment (prcr) [inline docs/one-click-integrations/bitbucket-integration.md:35]: This is more like a note than anything else. :spiral_notepad:

I realize that @claudiacarpinteiro probably adapted the naming of the screenshots from the GitHub integration ones. In the long run, it's better to have more explicit file names to avoid "second-guessing" and make maintenance easier:

```suggestion
    ![Installing the Pulse Bitbucket App](images/bitbucket-installing.png)

1.  Choose the workspace that you want to connect the integration to and click **Complete setup** to confirm. You can only select a workspace where you have owner permissions.

    ![Choosing the Bitbucket workspace](images/bitbucket-workspace.png)

1.  Wait until you get a confirmation that Pulse successfully connected the integration to Bitbucket.

    ![Pulse Bitbucket integration set up successfully](images/bitbucket-ok.png)
```

We should also update the names of the GitHub integration later when we refresh them. :+1:

Comment (nicklem) [reply docs/one-click-integrations/bitbucket-integration.md:35]: OK, updated

**Pull request merged**
